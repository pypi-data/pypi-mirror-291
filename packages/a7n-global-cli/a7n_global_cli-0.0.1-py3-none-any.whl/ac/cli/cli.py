import click
from xxxai.version import __version__

from . import deployment
from . import infer
from .util import click_group

CONTEXT_SETTINGS = dict(help_option_names=["-h", "--help"])


@click.version_option(__version__, "-v", "--version")
@click_group(context_settings=CONTEXT_SETTINGS)
def main():
    pass


deployment.add_command(main)
infer.add_command(main)
