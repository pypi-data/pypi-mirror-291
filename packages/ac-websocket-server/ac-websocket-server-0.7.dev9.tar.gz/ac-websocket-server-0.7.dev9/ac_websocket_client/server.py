#!/usr/bin/env python

'''Server UI'''

import tkinter as tk

from ac_websocket_client.objects import (
    GriddedButton, GriddedEntry, GriddedFrame, GriddedLabel, TrafficLight)


class ServerUI(GriddedFrame):
    '''Server UI'''

    def __init__(self, parent):

        super().__init__(grid_row=1, grid_col=0, height_by=3.5)

        self.parent = parent

        self.configure_columns(1, 1, 4, 1, 1)

        self._buttons = {}
        self._fields = {}
        self._lights = {}

        grid_row = 0

        GriddedLabel(self, grid_row=grid_row, grid_col=0,
                     width=8, text="Game")

        GriddedLabel(self, grid_row=grid_row, grid_col=1,
                     width=8, text="started:")
        self._fields['started'] = tk.StringVar()
        GriddedEntry(self, grid_row=grid_row, grid_col=2,
                     textvariable=self._fields['started'], state=tk.DISABLED)
        self._buttons['game'] = tk.StringVar(value='Start Game')
        GriddedButton(self, grid_row=grid_row, grid_col=3,
                      textvariable=self._buttons['game'],
                      command=lambda: self.parent.loop.create_task(self.parent.toggle_game()))
        self._lights['game'] = TrafficLight(self, row=grid_row, column=4)

        grid_row += 1

        self._fields['registered'] = tk.StringVar()
        GriddedLabel(self, grid_row=grid_row, grid_col=1,
                     width=8, text="registered:")
        GriddedEntry(self, grid_row=grid_row, grid_col=2,
                     textvariable=self._fields['registered'], state=tk.DISABLED)
        self._lights['registered'] = TrafficLight(self, row=grid_row, column=4)
        self._buttons['lobby'] = tk.StringVar(value='(Re)register')
        GriddedButton(self, grid_row=grid_row, grid_col=3,
                      textvariable=self._buttons['lobby'],
                      command=lambda: self.parent.loop.create_task(self.parent.toggle_registration()))
        self._lights['lobby'] = TrafficLight(self, row=grid_row, column=4)

        grid_row += 1

        GriddedLabel(self, grid_row=grid_row,
                     grid_col=0, width=8, text="Config")

        self._fields['cfg'] = tk.StringVar()
        GriddedLabel(self, grid_row=grid_row, grid_col=1, width=8, text="cfg:")
        GriddedEntry(self, grid_row=grid_row, grid_col=2,
                     textvariable=self._fields['cfg'], state=tk.DISABLED)
        self._buttons['save'] = tk.StringVar(value='Save .cfg')
        GriddedButton(self, grid_row=grid_row, grid_col=3,
                      textvariable=self._buttons['save'],
                      command=lambda: self.parent.loop.create_task(self.parent.save_configuration()))
        self._lights['save'] = TrafficLight(self, row=grid_row, column=4)

        grid_row += 1

        self._fields['name'] = tk.StringVar()
        GriddedLabel(self, grid_row=grid_row,
                     grid_col=1, width=8, text="name:")
        GriddedEntry(self, grid_row=grid_row, grid_col=2,
                     textvariable=self._fields['name'], state=tk.DISABLED)

        grid_row += 1

        self._fields['track'] = tk.StringVar()
        GriddedLabel(self, grid_row=grid_row,
                     grid_col=1, width=8, text="track:")
        GriddedEntry(self, grid_row=grid_row, grid_col=2,
                     textvariable=self._fields['track'], state=tk.DISABLED)

        grid_row += 1

        self._fields['cars'] = tk.StringVar()
        GriddedLabel(self, grid_row=grid_row,
                     grid_col=1, width=8, text="cars:")
        GriddedEntry(self, grid_row=grid_row, grid_col=2,
                     textvariable=self._fields['cars'], state=tk.DISABLED)

        self._buttons['sessions'] = {}
        self._fields['sessions'] = {}
        self._lights['sessions'] = {}

        grid_row += 1

        self._fields['sessions']['Practice'] = tk.StringVar()
        GriddedLabel(self, grid_row=grid_row, grid_col=1,
                     width=8, text="practice:")
        GriddedEntry(self, grid_row=grid_row, grid_col=2,
                     textvariable=self._fields['sessions']['Practice'], state=tk.DISABLED)
        self._buttons['sessions']['Practice'] = tk.StringVar(value='Enable')
        GriddedButton(self, grid_row=grid_row, grid_col=3,
                      textvariable=self._buttons['sessions']['Practice'],
                      command=lambda: self.parent.loop.create_task(self.parent.toggle_session('Practice')))
        self._lights['sessions']['Practice'] = TrafficLight(
            self, row=grid_row, column=4)

        grid_row += 1

        self._fields['sessions']['Qualify'] = tk.StringVar()
        GriddedLabel(self, grid_row=grid_row, grid_col=1,
                     width=8, text="qualify:")
        GriddedEntry(self, grid_row=grid_row, grid_col=2,
                     textvariable=self._fields['sessions']['Qualify'], state=tk.DISABLED)
        self._buttons['sessions']['Qualify'] = tk.StringVar(value='Enable')
        GriddedButton(self, grid_row=grid_row, grid_col=3,
                      textvariable=self._buttons['sessions']['Qualify'],
                      command=lambda: self.parent.loop.create_task(self.parent.toggle_session('Qualify')))
        self._lights['sessions']['Qualify'] = TrafficLight(
            self, row=grid_row, column=4)

        grid_row += 1

        self._fields['sessions']['Race'] = tk.StringVar()
        GriddedLabel(self, grid_row=grid_row, grid_col=1,
                     width=8, text="race:")
        GriddedEntry(self, grid_row=grid_row, grid_col=2,
                     textvariable=self._fields['sessions']['Race'], state=tk.DISABLED)
        self._buttons['sessions']['Race'] = tk.StringVar(value='Enable')
        GriddedButton(self, grid_row=grid_row, grid_col=3,
                      textvariable=self._buttons['sessions']['Race'],
                      command=lambda: self.parent.loop.create_task(self.parent.toggle_session('Race')))
        self._lights['sessions']['Race'] = TrafficLight(
            self, row=grid_row, column=4)

        self.update_ui()

    def update_ui(self):
        '''Update the UI with the contents of the parent.server'''

        self._fields['cfg'].set(self.parent.server.get('child_ini_file', ''))
        self._fields['name'].set(self.parent.server.get('name', ''))
        self._fields['track'].set(self.parent.server.get('track', ''))
        self._fields['cars'].set(self.parent.server.get('cars', ''))

        if not self.parent.states.is_connected:
            self._lights['game'].gray()
            self._lights['lobby'].gray()
            for session in ('Practice', 'Qualify', 'Race'):
                if self.parent.sessions.get(session, None):
                    self.parent.sessions[session]['active'] = False
                    self._lights['sessions'][session].gray()

        if self.parent.states.is_started:
            self._fields['started'].set(
                self.parent.server.get('timestamp', None))
            self._buttons['game'].set('Stop Game')
            self._lights['game'].green()
        else:
            self._fields['started'].set('n/a')
            self._buttons['game'].set('Start Game')
            self._lights['game'].red()

        if self.parent.states.is_registered and self.parent.states.is_started:
            self._fields['registered'].set(
                self.parent.lobby.get('since', ''))
            self._buttons['lobby'].set('Re-register')
            self._lights['lobby'].green()
        else:
            self._fields['registered'].set('n/a')
            self._buttons['lobby'].set('Register')
            self._lights['lobby'].red()

        if self.parent.states.cfg_needs_saving:
            self._lights['save'].red()
        else:
            self._lights['save'].gray()

        if sessions := self.parent.sessions:
            for session_type in sessions:
                if session_type not in ('Practice', 'Qualify', 'Race'):
                    return
                self._buttons['sessions'][session_type].set('Disable')
                session_active = sessions[session_type]['active']
                if session_active:
                    self._lights['sessions'][session_type].green()
                else:
                    self._lights['sessions'][session_type].gray()
                session_description = str(
                    sessions[session_type]['time'])
                if sessions[session_type]['laps'] == 0:
                    session_description += ' minutes'
                else:
                    session_description += ' laps'
                self._fields['sessions'][session_type].set(session_description)
