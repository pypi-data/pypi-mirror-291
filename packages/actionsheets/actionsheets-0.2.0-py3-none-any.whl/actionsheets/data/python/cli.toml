language = "python"
parent = "python"
name = "cli"
title = "Command-line interface"
description = """
Defining and processing a command-line interface.
Uses the [argparse](https://docs.python.org/3/library/argparse.html#) module.
"""

[create]
section = "Create"
description = "Create an command-line arguments parser"

[create.argparse]
action = "Define an argument parser"
code = """
import argparse
parser = argparse.ArgumentParser(
  prog = 'My CLI program',
  description = 'Description of functionality here',
  epilog = 'Text at the bottom of the help file'
)
"""

[create.argparse.add]
section = "Add argument to the parser"

[create.argparse.add.pos]
action = "Define positional argument"
code = "parser.add_argument('file')"

[create.argparse.add.pos.optional]
action = "Define optional positional argument"
code = "parser.add_argument('dest', nargs='?')"

[create.argparse.add.pos.list]
action = "Define positional argument accepting multiple values"
code = "parser.add_argument('files', type = str, nargs='+')"

[create.argparse.add.flag.enable]
action = "Define flag to enable a feature (default is False)"
code = "parser.add_argument('-f', '--force', action = 'store_true')"

[create.argparse.add.flag.disable]
action = "Define flag to disable a feature (default is True)"
code = "parser.add_argument('--disable', action = 'store_false')"

[create.argparse.add.option.str]
action = "Define string option"
code = "parser.add_argument('-s', '--source', type = str)"

[create.argparse.add.option.int]
action = "Define int option"
code = "parser.add_argument('--seed', type = int)"

[create.argparse.add.option.float]
action = "Define float option"
code = "parser.add_argument('value', type = float)"

[create.argparse.add.option.cat]
action = "Define categorical option"
code = "parser.add_argument('answer', choices = ['a', 'b', 'c'])"

[create.argparse.add.option.path]
action = "Define valid path option"
code = "parser.add_argument('path', type = pathlib.Path)"

[create.argparse.add.option.file.open]
action = "Define openable file option"
code = "parser.add_argument('file', type = open)"

[create.argparse.add.option.file.write]
action = "Define writable file option"
code = "parser.add_argument('dest', type = argparse.FileType('w'))"

[create.argparse.add.option.value]
action = "Define option that takes a value"
code = "parser.add_argument('--seed')"

[create.argparse.add.option.value.optional]
action = "Define option that optionally takes a value"
code = "parser.add_argument('--seed', nargs = '?')"

[create.argparse.add.option.required]
action = "Define required option"
code = "parser.add_argument('--seed', required = True)"
details = "Required options are considered bad form. Use positional arguments instead."


[parse]
section = "Parse"

[parse.sysargs]
action = "Parse system arguments using the created argparse instance"
code = "args = parser.parse_args()"


[test]
section = "Test"

[test.narg]
action = "Test if no arguments were provided"
code = "len(sys.argv) <= 1"


[get]
section = "Get"

[get.invoke]
action = "Get program or script that invoked the process"
code = "sys.argv[0]"

[get.nargs]
action = "Number of arguments"
code = "len(sys.argv)"
