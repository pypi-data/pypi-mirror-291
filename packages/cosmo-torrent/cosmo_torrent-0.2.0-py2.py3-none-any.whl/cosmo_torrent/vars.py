import os

COSMO_TORRENT_SSH_SERVER = os.environ.get(
    "COSMO_TORRENT_SSH_SERVER", "login.phys.ethz.ch"
)
COSMO_TORRENT_SSH_PORT = os.environ.get("COSMO_TORRENT_SSH_PORT", "22")

COSMO_TORRENT_REMOTE_FOLDER = os.environ.get(
    "COSMO_TORRENT_REMOTE_FOLDER",
    "/net/ipa-gate.phys.ethz.ch/export/ipa/public_html/cosmo-torrent",
)

COSMO_TORRENT_BASE_URL = os.environ.get(
    "COSMO_TORRENT_BASE_URL", "https://share.phys.ethz.ch/~ipa/cosmo-torrent"
)

MARKER_FILE = ".valid"
