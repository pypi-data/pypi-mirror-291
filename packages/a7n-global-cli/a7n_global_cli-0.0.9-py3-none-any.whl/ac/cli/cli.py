import click
from ac.version import __version__

from ac.cli import deployment
from ac.cli import dataset
from ac.cli.util import click_group

CONTEXT_SETTINGS = dict(help_option_names=["-h", "--help"])


@click.version_option(__version__, "-v", "--version")
@click_group(context_settings=CONTEXT_SETTINGS)
def main():
    pass


deployment.add_command(main)
dataset.add_command(main)

if __name__ == "__main__":
    main()