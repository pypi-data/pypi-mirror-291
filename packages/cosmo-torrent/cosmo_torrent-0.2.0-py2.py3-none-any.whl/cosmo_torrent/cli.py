import os
import shutil
import sys
import tempfile

import click
import requests

from .cosmo_torrent import _md5_sum, data_path
from .vars import (
    COSMO_TORRENT_BASE_URL,
    COSMO_TORRENT_REMOTE_FOLDER,
    COSMO_TORRENT_SSH_PORT,
    COSMO_TORRENT_SSH_SERVER,
)


@click.group()
def cosmo_torrent():
    pass


@cosmo_torrent.command()
@click.argument("identifier")
def download(identifier):
    sys.exit(_download(identifier))


def _download(identifier):
    try:
        folder = data_path(identifier)
    except Exception as e:
        click.secho(f"error: {e}", fg="red")
        return 1
    click.secho("data available at", fg="green")
    click.secho(f"{folder}", fg="green")
    return 0


@cosmo_torrent.command()
@click.option(
    "--identifier",
    help="data set identifier for later download",
    type=str,
    required=True,
)
@click.option(
    "--user",
    help="user name for scp to server",
    default=os.environ.get("USER"),
    show_default=True,
    type=str,
)
@click.option(
    "--force",
    help="overwrite existing dataset",
    is_flag=True,
)
@click.argument("folder")
def upload(identifier, user, folder, force):
    sys.exit(_upload(identifier, user, folder, force))


def _upload(
    identifier,
    user,
    folder,
    force,
    COSMO_TORRENT_SSH_SERVER=COSMO_TORRENT_SSH_SERVER,
    COSMO_TORRENT_SSH_PORT=COSMO_TORRENT_SSH_PORT,
    COSMO_TORRENT_REMOTE_FOLDER=COSMO_TORRENT_REMOTE_FOLDER,
):
    if not os.path.exists(folder):
        click.secho(f"folder {folder} does not exist", fg="red")
        return 1

    existing = _list()
    if identifier in existing and not force:
        click.secho(
            "data set with identifier {identifier} exists already."
            " use --force to overwrite.",
            fg="red",
        )
        return 1

    prep_folder = tempfile.mkdtemp()
    click.secho("create archive", fg="green")
    shutil.make_archive(os.path.join(prep_folder, identifier), "zip", folder, ".")
    md5_sum = _md5_sum(os.path.join(prep_folder, identifier) + ".zip")
    click.secho("create md5 file", fg="green")
    with open(os.path.join(prep_folder, identifier) + ".md5", "w") as fh:
        fh.write(md5_sum)

    result = os.system(
        f"scp -P {COSMO_TORRENT_SSH_PORT} -o StrictHostKeyChecking=no"
        f" {prep_folder}/*"
        f" {user}@{COSMO_TORRENT_SSH_SERVER}:{COSMO_TORRENT_REMOTE_FOLDER}"
    )

    if result == 0:
        click.secho(f"uploaded files from {folder} to server", fg="green")
    else:
        click.secho(f"upload of {folder} to server failed", fg="red")
    return result


def _list(url=COSMO_TORRENT_BASE_URL):
    response = requests.get(url)
    from bs4 import BeautifulSoup

    soup = BeautifulSoup(response.text, "html.parser")
    links = soup.find_all("a")
    hrefs = set(l.get("href") for l in links)
    stems = {href[:-4] for href in hrefs if href.endswith(".md5")}
    valid = sorted(stem for stem in stems if stem + ".zip" in hrefs)
    return valid


@cosmo_torrent.command()
def list():
    identifiers = _list()
    for id_ in identifiers:
        click.secho(id_, fg="blue")
