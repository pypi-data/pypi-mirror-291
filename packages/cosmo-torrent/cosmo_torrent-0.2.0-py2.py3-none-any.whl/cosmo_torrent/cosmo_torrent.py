# Copyright (C) 2024 ETH Zurich, Institute for Particle Physics and Astrophysics

import hashlib
import os
import sys
import tempfile
import zipfile

import requests
import tqdm
from requests.exceptions import HTTPError

from .vars import COSMO_TORRENT_BASE_URL, MARKER_FILE


def data_path(identifier):
    cache_folder = local_cache_folder(identifier)
    if not is_valid(cache_folder):
        download_data(identifier, cache_folder)
    return cache_folder


def _md5_sum(path, CHUNK_SIZE=512 * 1024**2):
    digest = hashlib.md5()
    with open(path, "rb") as fh:
        while True:
            chunk = fh.read(CHUNK_SIZE)
            if not chunk:
                return digest.hexdigest()
            digest.update(chunk)


def local_cache_folder(identifier):
    return os.path.join(_cache_base_folder(), identifier)


def _cache_base_folder():
    home = os.path.expanduser("~")
    if sys.platform.startswith("darwin"):
        base_folder = os.path.join(home, "Library", "Cache", "cosmo-torrent")
    else:
        base_folder = os.path.join(home, "_cache", "cosmo-torrent")
    return base_folder


def is_valid(cache_folder):
    return os.path.exists(os.path.join(cache_folder, MARKER_FILE))


def download_data(
    identifier, cache_folder, COSMO_TORRENT_BASE_URL=COSMO_TORRENT_BASE_URL
):
    _download_and_check_data(identifier, cache_folder, COSMO_TORRENT_BASE_URL)
    with open(os.path.join(cache_folder, MARKER_FILE), "w"):
        pass


def download(url, target, show_progressbar):
    try:
        response = requests.get(url, stream=True)
        response.raise_for_status()
        total_size = int(response.headers.get("content-length", 0))
        block_size = 1024 * 1024
        with tqdm.tqdm(
            total=total_size,
            unit_scale=True,
            unit="B",
            unit_divisor=block_size,
            disable=not show_progressbar,
        ) as progress_bar:
            with open(target, "wb") as file:
                for data in response.iter_content(block_size):
                    progress_bar.update(len(data))
                    file.write(data)
    except HTTPError as e:
        if e.response.status_code == 404:
            raise IOError("data set does not exist")
        raise e


def _download_and_check_data(identifier, cache_folder, url):
    with tempfile.TemporaryDirectory() as download_folder:
        archive = identifier + ".zip"
        downloaded_archive = os.path.join(download_folder, archive)
        download(url.rstrip("/") + "/" + archive, downloaded_archive, True)

        md5 = identifier + ".md5"
        downloaded_md5 = os.path.join(download_folder, md5)
        download(url.rstrip("/") + "/" + md5, downloaded_md5, False)

        if check_downloaded_data(downloaded_archive, downloaded_md5):
            with zipfile.ZipFile(downloaded_archive, "r") as zh:
                zh.extractall(cache_folder)
            return cache_folder
    raise IOError("checksum failure after download")


def check_downloaded_data(downloaded_archive, downloaded_md5):
    with open(downloaded_md5, "r") as fh:
        md5sum = fh.read().strip()

    return _md5_sum(downloaded_archive) == md5sum


if __name__ == "__main__":
    print(data_path("ucat.test"))
