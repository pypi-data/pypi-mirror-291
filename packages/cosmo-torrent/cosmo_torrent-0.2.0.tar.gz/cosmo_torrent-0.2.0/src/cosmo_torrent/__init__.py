__author__ = "Uwe Schmitt"
__email__ = "uwe.schmitt@id.ethz.ch"
__credits__ = "ETH Zurich, Institute for Particle Physics and Astrophysics"

__version__ = "0.2.0"

from .cosmo_torrent import data_path  # noqa: F401
