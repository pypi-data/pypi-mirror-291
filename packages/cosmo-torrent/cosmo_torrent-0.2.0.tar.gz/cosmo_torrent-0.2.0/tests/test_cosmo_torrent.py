import os
import socket
import threading
from functools import partial
from http.server import HTTPServer, SimpleHTTPRequestHandler
from urllib.request import urlretrieve

from cosmo_torrent.cli import _list, _upload
from cosmo_torrent.cosmo_torrent import download_data


def _free_port():
    # https://stackoverflow.com/questions/1365265
    sock = socket.socket()
    sock.bind(("", 0))
    return sock.getsockname()[1]


def test_upload_and_download(tmpdir_factory, regtest):
    user = os.environ.get("USER")
    data_to_upload_folder = tmpdir_factory.mktemp("data")
    data_to_upload_folder.join("data.txt").write_text("abc", "utf-8")

    local_data_storage = tmpdir_factory.mktemp("storage").strpath

    assert (
        _upload(
            "test",
            user,
            data_to_upload_folder,
            False,
            "localhost",
            22,
            local_data_storage,
        )
        == 0
    )

    p = _free_port()

    try:
        server = HTTPServer(
            ("", p), partial(SimpleHTTPRequestHandler, directory=local_data_storage)
        )
        t = threading.Thread(target=server.serve_forever)
        t.start()

        url = f"http://localhost:{p}"
        assert _list(url) == ["test"]

        downloaded, _ = urlretrieve(url)
        print(open(downloaded).read(), file=regtest)

        download_folder = tmpdir_factory.mktemp("downloads")

        download_data("test", download_folder.strpath, url)
        for p in download_folder.listdir():
            print(p.strpath, file=regtest)

    finally:
        server.shutdown()
        t.join()
