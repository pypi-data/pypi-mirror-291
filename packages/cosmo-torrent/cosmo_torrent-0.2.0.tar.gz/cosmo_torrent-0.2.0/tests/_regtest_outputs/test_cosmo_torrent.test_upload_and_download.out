<!DOCTYPE HTML>
<html lang="en">
<head>
<meta charset="utf-8">
<title>Directory listing for /</title>
</head>
<body>
<h1>Directory listing for /</h1>
<hr>
<ul>
<li><a href="test.md5">test.md5</a></li>
<li><a href="test.zip">test.zip</a></li>
</ul>
<hr>
</body>
</html>

<pytest_tempdir>/downloads0/.valid
<pytest_tempdir>/downloads0/data.txt
